pkg/operator/status/status_controller.go

```diff
@@ -0,0 +1,407 @@
+		}
+		relatedObjects = append(relatedObjects, handlerRelatedObjects...)
+	}
+
+	// sets defaults for condition not set by any handler
+	conditions = defaultUnsetConditions(conditions)
+
+	// at this point we know all condition types exist in conditions
+
+	co.Status.Conditions = conditions
+	co.Status.RelatedObjects = relatedObjects
+	co.Status.Versions = computeClusterOperatorVersions()
+
+	// check if the operator is disabled and reflect that in the Available condition
+	mode, _, err := utils.GetOperatorConfiguration(kubeClient, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if operator is disabled")
+	}
+	if mode == operatorv1.CloudCredentialsModeManual {
+		available, _ := findClusterOperatorCondition(co.Status.Conditions, configv1.OperatorAvailable)
+		if available.Status == configv1.ConditionTrue {
+			available.Reason = reasonOperatorDisabled
+			available.Message = msgOperatorDisabled
+		}
+	}
+
+	// Update transition time for any condition that has changed
+	setLastTransitionTime(oldConditions, co.Status.Conditions)
+
+	// Check if version changed, if so force a progressing last transition update:
+	if !reflect.DeepEqual(oldVersions, co.Status.Versions) {
+		logger.WithFields(log.Fields{
+			"old": oldVersions,
+			"new": co.Status.Versions,
+		}).Info("version has changed, updating progressing condition lastTransitionTime")
+		progressing, _ := findClusterOperatorCondition(co.Status.Conditions, configv1.OperatorProgressing)
+		// We know this should be there.
+		progressing.LastTransitionTime = metav1.Now()
+	}
+
+	// ClusterOperator should already exist (from the manifest payload), but recreate it if needed
+	if isNotFound {
+		if err := kubeClient.Create(context.TODO(), co); err != nil {
+			return errors.Wrap(err, "failed to create clusteroperator")
+		}
+		logger.Info("created clusteroperator")
+		// return error so we can immediately recalculate status???
+		return nil
+	}
+
+	// Update status fields if needed
+	if !reflect.DeepEqual(oldConditions, co.Status.Conditions) ||
+		!reflect.DeepEqual(oldVersions, co.Status.Versions) ||
+		!reflect.DeepEqual(oldRelatedObjects, co.Status.RelatedObjects) {
+
+		if err := kubeClient.Status().Update(context.TODO(), co); err != nil {
+			return errors.Wrap(err, "failed to update clusteroperator status")
+		}
+		logger.Info("clusteroperator status updated")
+	}
+
+	return nil
+}
+
+func mergeConditions(existing, new []configv1.ClusterOperatorStatusCondition, handlerName string, logger log.FieldLogger) []configv1.ClusterOperatorStatusCondition {
+	for _, newCondition := range new {
+		existingCondition, index := findClusterOperatorCondition(existing, newCondition.Type)
+		if existingCondition == nil {
+			existing = append(existing, newCondition)
+		} else {
+			logger.WithField("statushandler", handlerName).Infof("condition already set for type %s by a previous handler, the new condition from the current handler will be accepted: %v", newCondition.Type, newCondition)
```

Comment:
Last write wins? Could you clarify this in the function comment a little. This should probably be a warning as well as info is being lost. The suggestion above for logging every condition should help when combined with that.